### PR TITLE
WIP: Sync out latest changes and upgrade build process

### PR DESCRIPTION
**Pull request by zpao**
Goal is to get FDT to a sane place so it's much easier to maintain and use shipit. Obviously we've done a poor job of maintaining it. Ultimately it might even be best to sunset this and redirect the people to community driven projects in the same space.

The new build process is mostly just what we're doing in other projects. We can replace a bunch of 1-off code with shared code.

Most of the changes here are removal of code that will shift into fbjs (https://github.com/facebook/fbjs/pull/150). And then a bunch of var -> let/const that were part of a codemod. Not much actual code has changed and it's unclear if any PRs were merged here but not applied internally. Maybe @pieterv can spare a moment when this is closer to done to do a once over.

Tasks:
- [x] Update to latest internal code
- [x] Identify any external changes that should be retained
- [x] Update build process to use gulp
- [x] Add lint
- [x] ~~Add flow~~ (no flow in use right now, can add when/if annotations are added)
- [ ] Update website build process (currently too tightly coupled to raw untransformed code, breaks with use of fbjs)
- [x] Update release process to use new build process

### CODE REVIEW

File: src/FixedDataTableColumnNew.react.js [hunk 1]
```diff
@@ -151,7 +151,7 @@ var FixedDataTableColumn = React.createClass({
      * has a flex grow, once you resize the column this will be set to 0.
      *
      * This property only provides the UI for the column resizing. If this
-     * is set to true, you will need to set the onColumnResizeEndCallback table
+     * is set to true, you will need ot se the onColumnResizeEndCallback table
```
Comment (Daniel15) [inline src/FixedDataTableColumnNew.react.js]: :confused:

**Comment by Daniel15** [comment]
Thanks for working on this! :smile:

File: src/FixedDataTableColumnNew.react.js (hunk 1 above)
Comment (zpao) [inline src/FixedDataTableColumnNew.react.js]: Yea, downside of not immediately syncing back in. #267 fixed it but never made it internally. We should probably do a quick pass over commits here to make sure nothing else was missed.

File: package.json [hunk 2]
```diff
@@ -4,33 +4,49 @@
   "description": "A React table component designed to allow presenting thousands of rows of data.",
   "main": "main.js",
   "peerDependencies": {
-    "react": ">=0.13.0 <0.15.0 || ^0.14.0-beta3",
-    "react-dom": ">=0.14.0 <0.15.0 || ^0.14.0-beta3"
+    "react": "^0.15.0",
+    "react-dom": "^0.15.0"
```
Comment (pieterv) [inline package.json]: These should be `^15.0.0` no?

File: scripts/postcss/fbcss.js [hunk 3]
```diff
@@ -0,0 +1,32 @@
...
+    css.walkDecls(decl => {
+      const matches = decl.value.match(/var\((.+)\)/);
+      if (matches) {
+        if (matches[1] in colors) {
+          decl.value = colors[matches[1]];
+        } else {
+          throw new Error(`Missing ${matches[1]} from color mapping.`);
+        }
+      }
+    });
```
Comment (pieterv) [inline scripts/postcss/fbcss.js]: This is awesome, we can delete the runtime hack i added to cx :D

**Comment by pieterv** [comment]
Thanks for working on this @zpao. The synced code looks good, it contained everything I expected. The build system is looks nice although i didn't look into it with a huge amount of detail. Let me know if you want me to do another review or help out with testing.

**Comment by 3cooper** [comment]
Will this include any issue fixes or merges from other PR's?

**Comment by zpao** [comment]
> Will this include any issue fixes or merges from other PR's?

No, this is purely to get this building and syncing better with the tooling we have. The handful of code changes that are a part of this were just copied out from the internal version. Some open PRs could probably get closed as a result of better use of shared dependencies (eg #387).

**Comment by zpao** [comment]
Ok, everything is updated and actually builds now. I'm 99.99% certain it doesn't run properly (especially since `cssVar` is just going to throw right now: https://github.com/facebook/fbjs/pull/150/files#diff-5060a7b6b38915d3ac82bf25f1a2e74cR11). I'm still thinking on the best way to actually make that work. I don't want to duplicate the colors list, which currently we have partially in here, so perhaps I'll move it up into a standalone package in fbjs and we can have fbjs & the css transform consume it…

**Comment by KamranAsif** [comment]
Whats the motivation to move towards gulp? We've switched to babel 6 in our fork 
@wcjordan I think there are some commits here we may want to cherry pick. Can we create a jira ticket to follow up on this next sprint?

File: src/FixedDataTableBufferedRows.react.js [hunk 4]
```diff
@@ -24,24 +24,25 @@ var {PropTypes} = React;
 var FixedDataTableBufferedRows = React.createClass({
 
   propTypes: {
-    isScrolling: PropTypes.bool,
+    bufferRowCount: PropTypes.number,
```
Comment (KamranAsif) [inline src/FixedDataTableBufferedRows.react.js]: Was this changed somewhere before?

**Comment by zpao** [comment]
Motivation is that we have a bunch of infra built around gulp that we're using for our other projects, making it much easier to contribute for people at FB (especially, but hopefully also people in open source). Using that shared infra also means that we don't have 7 different ways of doing the same thing, each slightly differently (eg converting providesModule to commonjs).

File: src/FixedDataTableBufferedRows.react.js (hunk 4 above)
Comment (zpao) [inline src/FixedDataTableBufferedRows.react.js]: Internally. This is just a blind sync of the files as they exist today.

Huge motivation for this is to move to tooling that will sync every internal change out within minutes and make it painless for us to bring pull requests in.

**Comment by zpao** [comment]
Alright, I *think* this is just about done. I tested the built package with both the dist and npm builds (via browserify).

There's still a bit of followup work to do but it's pretty much ready to start using our shipit tools. The website doesn't build at all (need to learn more webpack voodoo) but the code builds.  There's sti

FWIW, while the actual code here hasn't changed too much, we might want to clean up a little bit and then ship it as 0.7. The biggest immediate change is that CSS class names have changed (using hyphen instead of underscore). The upgrade to Babel 6 means the build is pretty different bit-wise, though functionally it should be pretty similar.

**Comment by zpao** [comment]
I clearly never did this and just archived the project. Oh well.

**Pull request closed without merging**